src/audio/pipewire.c

```diff
@@ -0,0 +1,419 @@
+    {
+        if (state->is_running == false)
+        {
+            message(4, "stream is paused, starting");
+            state->is_running = true;
+            pw_stream_set_active(state->stream, true);
+        }
+    }
+    pw_thread_loop_unlock(state->loop);
+    track_buffer_size = track.num_samples * state->stride; // we want bytes, not frames
+    // as long as we have samples in the buffer sent by speech dispatcher, we loop, blocking spd from calling us till the audio system played what we loaded in it
+    while (track_buffer_size > 0)
+    {
+        message(5, "still have samples in the spd buffer");
+        // drainage spinlock!
+        // we wait till fill_quantity is not empty after this thread can continue, aka after on_process updated that ringbuffer index with what it managed to play, so that we have some room in the buffer to fill from what's remaining of what was given to us by speech dispatcher
+        while (true)
+        {
+            // we generally spend a lot of time waiting for this to happen, enough so that there can be something interrupting us, for example, speech dispatcher deciding that another piece of audio must have priority, while we're blocked in here. So, we check, each iteration of this loop, if  we were stopped by something, upon which we imediately return
+            if (state->is_running == false)
+                return 0;
+            // how much of the buffer was actually filled and not played yet by on_process
+            fill_quantity = spa_ringbuffer_get_write_index(&state->rb, &write_index);
+            // see if there's some room in the buffer, if so, stop locking
+            room_left_in_buffer = SAMPLE_BUFFER_SIZE - fill_quantity;
+            if (room_left_in_buffer > 0)
+            {
+                // then we have room in the buffer, so stop locking our thread
+                message(4, "we have %i bytes of free room in buffer, stop locking", room_left_in_buffer);
+                break;
+            }
+            // otherwise, to not starve a cpu core of resources, we suspend our spinning until something, in this case on_process, wakes us up
+            message(4, "no space, waiting for wakeup");
+            spa_system_eventfd_read(state->inner_loop->system, state->eventfd_number, &dummy);
+        }
+        // we write the amount of samples we can at this time without overflowing the buffer, to the memory area represented by the amount of room that's free after the last on_process call, to then be enqueued by pipewire
+        //  we assume speech dispatcher gives us the correct number of bytes for the format it chose, enough for this chunk. If that's not the case, there's not much we could do besides reading uninitialized memory or an incomplete chunk, unfortunately
+        // in case we produced more audio than the available room in the buffer that got freed by on_process, we only push what we can, otherwise we would read from uninitialised memory
```

Comment:
The comment is not coherent: when there is more audio to push than available room, we wouldn't be reading unitialized memory, but overwriting audio being played

Reply:
You probably rather meant: when there is less audio to push than available room

Reply:
yep, that's what I ment indeed, sorry for the confusion